hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/filter/ReclaimableKeyFilter.java

```diff
@@ -0,0 +1,161 @@
+  }
+
+  @Override
+  protected String getBucketName(Table.KeyValue<String, OmKeyInfo> keyValue) throws IOException {
+    return keyValue.getValue().getBucketName();
+  }
+
+  @Override
+  protected Boolean isReclaimable(Table.KeyValue<String, OmKeyInfo> deletedKeyInfo) throws IOException {
+    ReferenceCounted<OmSnapshot> previousSnapshot = getPreviousOmSnapshot(1);
+
+
+    KeyManager previousKeyManager = Optional.ofNullable(previousSnapshot)
+        .map(i -> i.get().getKeyManager()).orElse(null);
+
+
+    // Getting keyInfo from prev snapshot's keyTable/fileTable
+    CheckedSupplier<Optional<OmKeyInfo>, IOException> previousKeyInfo =
+        MemoizedCheckedSupplier.valueOf(() -> getPreviousSnapshotKey(getVolumeId(), getBucketInfo(),
+            deletedKeyInfo.getValue(), getKeyManager(), previousKeyManager));
+    // If file not present in previous snapshot then it won't be present in previous to previous snapshot either.
+    if (!previousKeyInfo.get().isPresent()) {
+      return true;
+    }
+
+    ReferenceCounted<OmSnapshot> previousToPreviousSnapshot = getPreviousOmSnapshot(0);
+    KeyManager previousToPreviousKeyManager = Optional.ofNullable(previousToPreviousSnapshot)
+        .map(i -> i.get().getKeyManager()).orElse(null);
+
+    // Getting keyInfo from prev to prev snapshot's keyTable/fileTable based on keyInfo of prev keyTable
+    CheckedSupplier<Optional<OmKeyInfo>, IOException> previousPrevKeyInfo =
+        MemoizedCheckedSupplier.valueOf(() -> getPreviousSnapshotKey(
+            getVolumeId(), getBucketInfo(), previousKeyInfo.get().orElse(null), previousKeyManager,
+            previousToPreviousKeyManager));
+    SnapshotInfo previousSnapshotInfo = getPreviousSnapshotInfo(1);
+    calculateExclusiveSize(previousSnapshotInfo, previousKeyInfo, previousPrevKeyInfo,
+        exclusiveSizeMap, exclusiveReplicatedSizeMap);
+    return false;
+  }
+
+  public Map<UUID, Long> getExclusiveSizeMap() {
+    return exclusiveSizeMap;
+  }
+
+  public Map<UUID, Long> getExclusiveReplicatedSizeMap() {
+    return exclusiveReplicatedSizeMap;
+  }
+
+  /**
+   * To calculate Exclusive Size for current snapshot, Check
+   * the next snapshot deletedTable if the deleted key is
+   * referenced in current snapshot and not referenced in the
+   * previous snapshot then that key is exclusive to the current
+   * snapshot. Here since we are only iterating through
+   * deletedTable we can check the previous and previous to
+   * previous snapshot to achieve the same.
+   * previousSnapshot - Snapshot for which exclusive size is
+   *                    getting calculating.
+   * currSnapshot - Snapshot's deletedTable is used to calculate
+   *                previousSnapshot snapshot's exclusive size.
+   * previousToPrevSnapshot - Snapshot which is used to check
+   *                 if key is exclusive to previousSnapshot.
+   */
+  private void calculateExclusiveSize(SnapshotInfo previousSnapshotInfo,
```

Comment:
does this fix the exclusive size bug? related: [HDDS-10051](https://issues.apache.org/jira/browse/HDDS-10051) https://issues.apache.org/jira/browse/HDDS-11528 https://issues.apache.org/jira/browse/HDDS-11151 https://issues.apache.org/jira/browse/HDDS-9431

Reply:
if so, shall we add checks in the test to verify exclusive size is expected

Reply:
No this doesn't fix that issue. Exclusive size calculation will be fixed as part of the key deleting service change

Reply:
https://github.com/apache/ozone/blob/bafa998896fc243adf09dc0bcd05136b5da5fb11/hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/filter/TestReclaimableKeyFilter.java#L212
There is a test case already here.